### PR TITLE
DSTORE-643 Add unit tests for SkipExistingDocumentsProcessorFactory.

### PR DESCRIPTION
**Pull request by timatbw**

### CODE REVIEW

File: solr/core/src/test/org/apache/solr/update/processor/SkipExistingDocumentsProcessorFactoryTest.java [hunk 1]
```diff
@@ -0,0 +1,328 @@
...
+import org.apache.solr.update.processor.SkipExistingDocumentsProcessorFactory.SkipExistingDocumentsUpdateProcessor;
+import org.junit.Test;
+import org.mockito.Mockito;
+
+public class SkipExistingDocumentsProcessorFactoryTest {
+
+  // Tests for logic in the factory
+
+  @Test(expected=SolrException.class)
+  public void testExceptionIfSkipInsertParamNonBoolean() throws Exception {
```
Comment (johnmbw) [inline solr/core/src/test/org/apache/solr/update/processor/SkipExistingDocumentsProcessorFactoryTest.java]: I don't think you need the `throws Exception` here? Don't think there are any methods in the test that declare they throw a checked exception.

File: solr/core/src/test/org/apache/solr/update/processor/SkipExistingDocumentsProcessorFactoryTest.java [hunk 2]
```diff
@@ -0,0 +1,328 @@
...
+    BytesRef docId = new BytesRef();
+    AddUpdateCommand cmd = new AddUpdateCommand(req);
+    cmd.setIndexedId(docId);
+    cmd.solrDoc = new SolrInputDocument();
+    assertFalse(AtomicUpdateDocumentMerger.isAtomicUpdate(cmd));
+    doReturn(true).when(processor).isLeader(cmd);
+    doReturn(true).when(processor).doesDocumentExist(docId);
+
+    processor.processAdd(cmd);
+    verify(next, times(1)).processAdd(cmd);
```
Comment (johnmbw) [inline solr/core/src/test/org/apache/solr/update/processor/SkipExistingDocumentsProcessorFactoryTest.java]: NABD `times(1)` not needed here

File: solr/core/src/test/org/apache/solr/update/processor/SkipExistingDocumentsProcessorFactoryTest.java [hunk 3]
```diff
@@ -0,0 +1,328 @@
...
+    SolrQueryRequest req = new LocalSolrQueryRequest(null, new NamedList());
+    UpdateRequestProcessor next = Mockito.mock(DistributedUpdateProcessor.class);
+    SkipExistingDocumentsUpdateProcessor processor
+            = Mockito.spy(new SkipExistingDocumentsUpdateProcessor(req, next, false, true));
+
+    BytesRef docId = new BytesRef();
+    AddUpdateCommand cmd = new AddUpdateCommand(req);
+    cmd.setIndexedId(docId);
+    cmd.solrDoc = new SolrInputDocument();
+    cmd.solrDoc.addField("last_name", ImmutableMap.of("set", "Smith"));
```
Comment (johnmbw) [inline solr/core/src/test/org/apache/solr/update/processor/SkipExistingDocumentsProcessorFactoryTest.java]: Maybe worth putting this bit into a `private` method (`createAtomicUpdateCmd`)? As it'll read better in the tests and avoid a bit of duplication.

File: solr/core/src/test/org/apache/solr/update/processor/SkipExistingDocumentsProcessorFactoryTest.java (hunk 1 above)
Comment (timatbw) [reply solr/core/src/test/org/apache/solr/update/processor/SkipExistingDocumentsProcessorFactoryTest.java]: Thanks - will tidy up these things you've highlighted. I borrowed some of the test outline from another Solr test class, so the `throws` might be an artifact of that.

Comment (johnmbw) [reply solr/core/src/test/org/apache/solr/update/processor/SkipExistingDocumentsProcessorFactoryTest.java]: Yeah I figured it was probably to match the solr code (was trying to restrain my pickiness a bit 😄 )

**Comment by timatbw** [comment]
Have updated, removed many redundant `throws` clauses, refined others to `IOException`, removed redundant `times(1)`, refactored some repeated code, and added 2 new tests to check the leader logic.

**Review by johnmbw — APPROVED** [review]
LGTM

**Pull request merged**